July-Morning/assignment1/Assignment_1_Q2.cpp

```diff
@@ -0,0 +1,104 @@
+#include <iostream>
+
+typedef struct node
+{
+    int data;
+    node * next;
+} node;
+
+class single_linked_list 
+{
+    private:
+        struct node * head;
+        
+    public:
+        single_linked_list() 
+        {
+            head = NULL;
+        }
+        
+        ~single_linked_list() 
+        {
+            while (head != NULL)
+            {
+                node * temp = head; 
+                head = head->next;
+                delete head;
+            }    
+        }
+
+        // Insert element with given data to the beginning of the list
+        void insert (int data) 
+        {
+            node * temp = new node;
+            temp->data = data;
+            if (head == NULL)
+            {
+                temp->next = NULL;
+                head = temp;
+            }
+            else
+            {
+                temp->next = head;
+                head = temp;
+            }    
+        }
+        
+        // Find k-th element from the end of the list
+        // Return value:
+        // true - if k-th element from the end was found successfully
+        // false - if size of the list is less than given k
+        bool find_k_from_tail (int k, int& data) 
+        {
+            node * tail = head;
+            node * k_node = head;
+            int diff_counter = 0;
+            while (tail != NULL)

```

Comment:
I understand that first you advance `diff_counter` until it gets to (k+1);
then you advance k_node.

This can be better expressed by breaking this to two loops:
```C++
while (tail != NULL && diff_counter < k + 1) {
    tail = tail->next;
    diff_counter++;
}
while (tail != NULL) {
    tail = tail->next;
    k_node = k_node->next;
}
```
This code may be longer but easier to follow.
It is easier to understand each loops purpuse:
First loop: Advance `tail` `k+1` times.
Second loop: Advance `tail` and `k_node` together until `tail` reaches the end.